### PR TITLE
*: update rust toolchain

### LINKED ISSUE #9444: make doc is broken
### What version of TiKV are you using?

master

### Steps to reproduce

`make doc`

### What did happened?

```
$ make doc
thread 'main' panicked at 'activated_features for invalid package: features did not find PackageId { name: "synstructure", version: "0.12.3", source: "registry `https://github.com/rust-lang/crates.io-index`" } false', src/tools/cargo/src/cargo/core/resolver/features.rs:227:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
make: *** [Makefile:277: doc] Error 101
```

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What problem does this PR solve?

Issue Number: close #9444

### What is changed and how it works?

update rust toolchain to nightly-2021-01-25

### Check List 

Tests 

- Unit test

### Release note 

No release note

### CODE REVIEW

**Comment by BusyJay** [comment]
Can you describe the motivation of updating it?

**Comment by tabokie** [comment]
@BusyJay Suggested by this [comment](https://github.com/tikv/tikv/pull/9458#issuecomment-765040194), the use of [`as_chunks`](https://github.com/tikv/tikv/pull/9458#issuecomment-765040194) method requires latest nightly toolchain. I'm not sure how we do the updates regularly, any hidden hazards here?

**Comment by BusyJay** [comment]
We usually update it every few months. I'm not objecting it, just curious if it's a routine update or for some specific functionality.

**Comment by sticnarf** [comment]
Could you wait for the 2021-01-25 toolchain for https://github.com/rust-lang/rust/pull/81245? With the latest fix on cargo (https://github.com/rust-lang/cargo/pull/9077), `cargo doc` will work for resolver v2

**Comment by tabokie** [comment]
@sticnarf Done.

File: components/cdc/src/service.rs [hunk 1]
```diff
@@ -414,18 +415,26 @@ mod tests {
...
+            ..Default::default()
+        };
 
-        let mut event_big = Event::default();
         let mut row_big = EventRow::default();
-        row_big.set_key(vec![0 as u8; CDC_MAX_RESP_SIZE as usize]);
-        let mut event_entries = EventEntries::default();
-        event_entries.entries = vec![row_big].into();
-        event_big.event = Some(Event_oneof_event::Entries(event_entries));
+        row_big.set_key(vec![0_u8; CDC_MAX_RESP_SIZE as usize]);
```
Comment (sticnarf) [inline components/cdc/src/service.rs:429]: Why doesn't clippy complain about it 🤣

File: components/test_raftstore/src/transport_simulate.rs [hunk 2]
```diff
@@ -720,10 +731,10 @@ impl Filter for LeadingDuplicatedSnapshotFilter {
...
 
     fn after(&self, res: Result<()>) -> Result<()> {
-        let dropped = self
-            .dropped
-            .compare_and_swap(true, false, Ordering::Relaxed);
-        if res.is_err() && dropped {
+        let dropped =
+            self.dropped
+                .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed);
+        if res.is_err() && dropped.is_ok() {
```
Comment (sticnarf) [inline components/test_raftstore/src/transport_simulate.rs:737]: I think moving `is_ok()` upward would be semantically better.

```suggestion
        let dropped =
            self.dropped
                .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
                .is_ok();
        if res.is_err() && dropped {
```

File: components/tidb_query_datatype/src/codec/data_type/mod.rs [hunk 3]
```diff
@@ -193,7 +193,7 @@ pub trait EvaluableRet: Clone + std::fmt::Debug + Send + Sync + 'static {
     type ChunkedType: ChunkedVec<Self>;
     /// Converts a vector of this concrete type into a `VectorValue` in the same type;
     /// panics if the varient mismatches.
-    fn into_vector_value(vec: Self::ChunkedType) -> VectorValue;
+    fn cast_into_vector_value(vec: Self::ChunkedType) -> VectorValue;
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/data_type/mod.rs:196]: The new name is still ambiguous. What about `cast_chunk_to_vector_value`?

File: components/tidb_query_datatype/src/codec/convert.rs [hunk 4]
```diff
@@ -591,13 +591,13 @@ pub fn bytes_to_uint_without_context(bytes: &[u8]) -> Result<u64> {
...
     if let Some(&c) = trimed.next() {
-        if c >= b'0' && c <= b'9' {
+        if (b'0'..=b'9').contains(&c) {
             r = Some(u64::from(c) - u64::from(b'0'));
         } else if c != b'+' {
             return Ok(0);
         }
 
-        for c in trimed.take_while(|&&c| c >= b'0' && c <= b'9') {
+        for c in trimed.take_while(|&&c| (b'0'..=b'9').contains(&c)) {
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/convert.rs:600]:
```suggestion
        for c in trimed.take_while(|&c| (b'0'..=b'9').contains(c)) {
```

File: components/tidb_query_datatype/src/codec/mysql/decimal.rs [hunk 5]
```diff
@@ -1849,7 +1849,7 @@ impl<'a> ConvertTo<Decimal> for JsonRef<'a> {
 fn first_non_digit(bs: &[u8], start_idx: usize) -> usize {
     bs.iter()
         .skip(start_idx)
-        .position(|&c| c < b'0' || c > b'9')
+        .position(|&c| !(b'0'..=b'9').contains(&c))
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/mysql/decimal.rs:1852]:
```suggestion
        .position(|c| !(b'0'..=b'9').contains(c))
```

File: components/tidb_query_datatype/src/codec/mysql/json/comparison.rs [hunk 6]
```diff
@@ -258,7 +258,7 @@ mod tests {
             let left: Json = left_str.parse().unwrap();
             let right: Json = right_str.parse().unwrap();
             assert!(left < right);
-            assert_eq!(left, left);
+            // assert_eq!(left, left);
```
Comment (sticnarf) [inline components/tidb_query_datatype/src/codec/mysql/json/comparison.rs:261]: Remove?

File: components/tidb_query_expr/src/impl_string.rs [hunk 7]
```diff
@@ -47,13 +47,13 @@ pub fn oct_string(s: BytesRef, writer: BytesWriter) -> Result<BytesGuard> {
...
             negative = true;
-        } else if c >= b'0' && c <= b'9' {
+        } else if (b'0'..=b'9').contains(&c) {
             r = Some(u64::from(c) - u64::from(b'0'));
         } else if c != b'+' {
             return Ok(writer.write(Some(b"0".to_vec())));
         }
 
-        for c in trimmed.take_while(|&&c| c >= b'0' && c <= b'9') {
+        for c in trimmed.take_while(|&&c| (b'0'..=b'9').contains(&c)) {
```
Comment (sticnarf) [inline components/tidb_query_expr/src/impl_string.rs:56]:
```suggestion
        for c in trimmed.take_while(|&c| (b'0'..=b'9').contains(c)) {
```

File: src/server/gc_worker/applied_lock_collector.rs [hunk 8]
```diff
@@ -339,21 +339,21 @@ impl LockCollectorRunner {
...
 
     fn stop_collecting(&mut self, max_ts: TimeStamp) -> Result<()> {
-        let curr_max_ts =
-            self.observer_state
-                .max_ts
-                .compare_and_swap(max_ts.into_inner(), 0, Ordering::SeqCst);
-        let curr_max_ts = TimeStamp::new(curr_max_ts);
-
-        if curr_max_ts == max_ts {
+        let curr_max_ts = self.observer_state.max_ts.compare_exchange(
```
Comment (sticnarf) [inline src/server/gc_worker/applied_lock_collector.rs:342]: Change the name `curr_max_ts` to something like `exchange_res`?

File: components/tidb_query_datatype/src/codec/data_type/mod.rs (hunk 3 above)
Comment (tabokie) [reply components/tidb_query_datatype/src/codec/data_type/mod.rs:196]: Sounds good

File: components/cdc/src/service.rs (hunk 1 above)
Comment (tabokie) [reply components/cdc/src/service.rs:429]: What do you mean?

**Review by sticnarf — COMMENTED** [review]
The rest lgtm

File: components/tidb_query_datatype/src/codec/mysql/json/comparison.rs (hunk 6 above)
Comment (tabokie) [reply components/tidb_query_datatype/src/codec/mysql/json/comparison.rs:261]: I'm not sure if the test writer intends to test the equal case, is it necessary to use `assert_eq!(left, left.clone())` instead?

Comment (sticnarf) [reply components/tidb_query_datatype/src/codec/mysql/json/comparison.rs:261]: Oh, kinda makes sense. I think clone sounds good, or we can just suppress the warning.

File: components/cdc/src/service.rs (hunk 1 above)
Comment (sticnarf) [reply components/cdc/src/service.rs:429]: I don't have a strong preference for the style. I just don't know why there isn't a warning about replacing it with:

```
EventRow {
   key: xxx,
   ..Default::default()
}
```

Comment (tabokie) [reply components/cdc/src/service.rs:429]: I guess clippy won't look into `set_key`'s detail, or check its name for that matter. I'll leave it be for now.

**Review by sticnarf — COMMENTED** [review]
/LGTM

**Review by sticnarf — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM. Though the `as_chunks` is not necessary now

**Pull request merged**